ros/src/tl_detector/tl_detector.py

```diff
@@ -113,39 +138,54 @@ def get_light_state(self, light):
             int: ID of traffic light color (specified in styx_msgs/TrafficLight)
 
         """
-        if(not self.has_image):
-            self.prev_light_loc = None
-            return False
-
-        cv_image = self.bridge.imgmsg_to_cv2(self.camera_image, "bgr8")
-
-        #Get classification
-        return self.light_classifier.get_classification(cv_image)
+        # TODO (Sergey Morozov): uncomment while classifier implemented
+        #if(not self.has_image):
+        #    self.prev_light_loc = None
+        #    return False
+        #
+        #cv_image = self.bridge.imgmsg_to_cv2(self.camera_image_msg, "bgr8")
+        #
+        ##Get classification
+        #return self.light_classifier.get_classification(cv_image)
+        return light.state
 
     def process_traffic_lights(self):
-        """Finds closest visible traffic light, if one exists, and determines its
-            location and color
+        """
+        Finds closest visible traffic light, if one exists, and determines its location and color.
 
         Returns:
             int: index of waypoint closes to the upcoming stop line for a traffic light (-1 if none exists)
             int: ID of traffic light color (specified in styx_msgs/TrafficLight)
-
         """
-        light = None
+        closest_light = None
+        line_wp_idx = None
 
-        # List of positions that correspond to the line to stop in front of for a given intersection
+        # List of positions that correspond to the line to stop in front of for a given intersection.
         stop_line_positions = self.config['stop_line_positions']
-        if(self.pose):
-            car_position = self.get_closest_waypoint(self.pose.pose)
 
-        #TODO find the closest visible traffic light (if one exists)
+        if self.pose_msg is not None:
+            car_wp_idx = self.get_closest_waypoint(self.pose_msg.pose.position.x,
+                                                   self.pose_msg.pose.position.y)
+
+            diff = len(self.waypoints_msg.waypoints)
+            for i, light in enumerate(self.lights):
+                # Get stop line waypoint index.
+                line = stop_line_positions[i]
+                temp_wp_idx = self.get_closest_waypoint(line[0], line[1])
+                # Find closest stop line waypoint index.
+                d = temp_wp_idx - car_wp_idx
+                if 0 <= d < diff:
+                    diff = d
```

Comment:
I think this code finds the closest traffic light, regardless of whether it is visible or not, this won't cause any harm here, because the waypoint updater ignores any traffic light outside the visible (LOOKAHEAD) range, but in real life, the car can't find traffic lights that are not visible.

Reply:
Yes, you are right. For SAE 5 this won't work. However, many autopilots rely on HD maps of the environment, where you actually know where the traffic light is based on the prior information. It is described in ["Self-Driving Cars Fundamentals: Featuring Apollo"](https://classroom.udacity.com/courses/ud0419) course from Udacity. So, the scenario we are working with here is applicable to the real world.